tests/fork-change-slot.c

```diff
@@ -0,0 +1,347 @@
+#define RANDOM_SIZE 20
+#define MAX_SIGSIZE 1024
+
+#if OPENSSL_VERSION_NUMBER < 0x10100003L
+#define EVP_PKEY_get0_RSA(key) ((key)->pkey.rsa)
+#endif
+
+static int do_wait(pid_t pids[], int num)
+{
+    int i;
+    int status = 0;
+
+    for (i = 0; i < num; i++) {
+        waitpid(pids[i], &status, 0);
+        if (WIFEXITED(status)) {
+            printf("child %d exited with status %d\n", pids[i], WEXITSTATUS(status));
+            return (WEXITSTATUS(status));
+        }
+        if (WIFSIGNALED(status)) {
+            fprintf(stderr, "Child %d terminated by signal #%d\n", pids[i],
+                    WTERMSIG(status));
+            return (WTERMSIG(status));
+        }
+        else {
+            perror("waitpid");
+        }
+    }
+
+    return 0;
+}
+
+static int spawn_processes(int num)
+{
+    int i;
+    int chld_ret = 0;
+    pid_t *pids;
+    pid_t pid;
+
+    sigset_t set, oldset;
+    int signal;
+
+    sigemptyset(&set);
+    sigaddset(&set, SIGUSR1);
+
+    /* If only 1 process was requested, no more processes are required */
+    if (num <= 1) {
+        return 0;
+    }
+
+    pids = (pid_t *)malloc(num * sizeof(pid_t));
+    if (pids == NULL) {
+        exit(ENOMEM);
+    }
+
+    /* Spawn (num - 1) new processes to get a total of num processes */
+    for (i = 0; i < (num - 1); i++) {
+        pid = fork();
+        switch (pid) {
+            case -1: /* failed */
+                perror("fork");
+                do_wait(pids, i);
+                free(pids);
+                exit(5);
+            case 0: /* child */
+                printf("Remove or add a device to cause an error\n");
+                printf("Waiting for signal SIGUSR1\n");
+                sigprocmask(SIG_BLOCK, &set, &oldset);
+                sigwait(&set, &signal);
+                sigprocmask(SIG_SETMASK, &oldset, NULL);
+                free(pids);
+                return 0;
+            default: /* parent */
+                pids[i] = pid;
+                printf("spawned %d\n", pid);
+        }
+    }
+
+    /* Wait for the created processes */
+    chld_ret = do_wait(pids, (num - 1));
+
+    free(pids);
+
+    return chld_ret;
+}
+
+static void error_queue(const char *name, int pid)
+{
+    if (ERR_peek_last_error()) {
+        fprintf(stderr, "pid %d: %s generated errors:\n", pid, name);
+        ERR_print_errors_fp(stderr);
+    }
+}
+
+static X509 *read_cert(char *file, int pid)
+{
+    BIO *bio = NULL;
+    X509 *cert = NULL; 
+
+    bio = BIO_new(BIO_s_file());
+
+    if (bio == NULL) {
+        error_queue("BIO_new", pid);
+        return NULL;
+    }
+
+    if (BIO_read_filename(bio, file) < 0) {
+        error_queue("BIO_read_filename", pid);
+        BIO_free(bio);
+        return NULL;
+    }
+
+    cert = PEM_read_bio_X509(bio, NULL, NULL, NULL);
+
+    if (cert == NULL) {
+        error_queue("PEM_read_bio_X509", pid);
+    }
+
+    BIO_free(bio);
+    return cert;
+}
+
+static void usage(char *argv[])
```

Comment:
no serious by why not only pass argv[0]?

Reply:
Previously I was using more arguments to write the usage, but then I changed the program. Since I am not using the other elements of the list any more, I'll change this accordingly.